File: src/test/java/com/esentri/wrabbitmq/java/SendReplyTest.java
package com.esentri.wrabbitmq.java;

import com.esentri.wrabbitmq.WrabbitTopic;
import org.junit.jupiter.api.Test;

import java.io.Serializable;
import java.util.UUID;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.IntStream;

import static com.esentri.wrabbitmq.java.TestFunctions.Await;
import static org.fest.assertions.Assertions.assertThat;

public class SendReplyTest {

   private final WrabbitTopic topic = new WrabbitTopic("TestTopic-Replier-Java");
   private <MESSAGE extends Serializable, REPLY extends Serializable> WrabbitEventWithReply<MESSAGE, REPLY> newEvent() {
      return new WrabbitEventWithReply<>(topic, UUID.randomUUID().toString());
   }

   @Test
   public void sendReplyStringToInt() {
      final WrabbitEventWithReply<String, Integer> event = newEvent();
      final CountDownLatch countDownLatch = new CountDownLatch(1);
      final int intValue = 12345;
      final String message = Integer.toString(intValue);

      event.replier(it -> {
         assertThat(it).isEqualTo(message);
         return Integer.parseInt(it);
      });
      event.sendAndReceive(message).thenAccept(it -> {
         assertThat(it).isEqualTo(intValue);
         countDownLatch.countDown();
      });

      Await(countDownLatch);
   }

   @Test
   public void sendReplyStringToInt_X_times() {
      final WrabbitEventWithReply<String, Integer> event = newEvent();
      final int intValue = 12345;
      final String message = Integer.toString(intValue);
      final int sentTimes = 100;
      final CountDownLatch countDownLatch = new CountDownLatch(sentTimes);

      TestDomain.ReplierTopic1.StringToInt.replier(it -> {
         assertThat(it).isEqualTo(message);
         return Integer.parseInt(it);
      });

      IntStream.range(0, sentTimes).forEach(it -> {
         TestDomain.ReplierTopic1.StringToInt.sendAndReceive(message).thenAccept(it2 -> {
            assertThat(it2).isEqualTo(intValue);
            countDownLatch.countDown();
         });
      });

      Await(countDownLatch);
   }

   @Test
   public void sendReplyStringToInt_2_replier_default() throws InterruptedException {
      final WrabbitEventWithReply<String, Integer> event = newEvent();
      // AtomicInteger is used to check after all expected events if any unexpected event occurs.
      // This is not possible with CountDownLatch.
      final AtomicInteger waitCounter = new AtomicInteger(0);
      final int intValue = 12345;
      final String message = Integer.toString(intValue);

      event.replier(it -> {
         assertThat(it).isEqualTo(message);
         return Integer.parseInt(it);
      });
      event.replier(it -> {
         assertThat(it).isEqualTo(message);
         return Integer.parseInt(it);
      });

      event.sendAndReceive(message).thenAccept(it -> {
         assertThat(it).isEqualTo(intValue);
         waitCounter.incrementAndGet();
      });

      while(waitCounter.get() <= 0) {
         Thread.sleep(300);
      }

      Thread.sleep(1000);
      assertThat(waitCounter.get()).isEqualTo(1);
   }

   @Test
   public void sendAndReplyWithContext() {
      final WrabbitEventWithReply<TestObjectObject, String> event = newEvent();
      final CountDownLatch countDownLatch = new CountDownLatch(1);
      final TestObjectObject message = new TestObjectObject(new TestObjectNumberText(12345, "hello world"));
      final String propertyKey = "test";
      final String propertyValue = "property";

      event.replier((context, it) -> {
         assertThat(it.getObj().getNumber()).isEqualTo(message.getObj().getNumber());
         assertThat(it.getObj().getText()).isEqualTo(message.getObj().getText());
         assertThat(context.get(propertyKey).toString()).isEqualToIgnoringCase(propertyValue);
         countDownLatch.countDown();
         return it.getObj().getText();
      });
      event
         .messageBuilder()
         .property(propertyKey, propertyValue)
         .sendAndReceive(message)
         .thenAccept(it -> assertThat(it).isEqualTo(message.getObj().getText()));

      Await(countDownLatch);
   }
}

File: src/test/kotlin/com/esentri/wrabbitmq/SendReplyTest.kt
package com.esentri.wrabbitmq

import org.fest.assertions.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.io.Serializable
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CountDownLatch
import java.util.concurrent.atomic.AtomicInteger

class SendReplyTest {

   private val topic = WrabbitTopic("TestTopic-Replier")
   private fun <MESSAGE : Serializable, REPLY : Serializable> newEvent() =
      NewEventWithReply<MESSAGE, REPLY>(topic)

   @Test
   fun sendReplyStringToInt() {
      val event = newEvent<String, Int>()
      val countDownLatch = CountDownLatch(1)
      val message = "12345"

      event.replier { it ->
         assertThat(it).isEqualTo(message)
         it.toInt()
      }
      event.sendAndReceive(message).thenAccept {
         assertThat(it).isEqualTo(message.toInt())
         countDownLatch.countDown()
      }

      Await(countDownLatch)
   }

   @Test
   fun sendReplyStringToInt_X_times() {
      val event = newEvent<String, Int>()
      val sentTimes = 100
      val countDownLatch = CountDownLatch(sentTimes)

      val message = "12345"
      event.replier { it ->
         assertThat(it).isEqualTo(message)
         it.toInt()
      }
      for (i in 1..sentTimes) {
         event.sendAndReceive(message).thenAccept {
            assertThat(it).isEqualTo(message.toInt())
            countDownLatch.countDown()
         }
      }

      Await(countDownLatch)
   }

   @Test
   fun sendReplyStringToInt_2_replier_default() {
      val event = newEvent<String, Int>()
      // AtomicInteger is used to check if any events occur after the expected.
      // This would not be possible with CountDownLatch.
      val waitCounter = AtomicInteger(0)
      val message = "12345"

      event.replier { it ->
         assertThat(it).isEqualTo(message)
         it.toInt()
      }
      event.replier { it ->
         assertThat(it).isEqualTo(message)
         it.toInt()
      }

      event.sendAndReceive(message).thenAccept {
         assertThat(it).isEqualTo(message.toInt())
         waitCounter.incrementAndGet()
      }

      while (waitCounter.get() <= 0) {
         Thread.sleep(300)
      }
      Thread.sleep(1000)
      assertThat(waitCounter.get()).isEqualTo(1)
   }

   @Test
   fun sendAndReplyWithContext() {
      val event = newEvent<TestObjectObject, String>()
      val countDownLatch = CountDownLatch(1)
      val message = TestObjectObject(TestObjectNumberText(12345, "hello world"))
      val propertyKey = "test"
      val propertyValue = "property"

      event.replier { context, it ->
         assertThat(it.obj.number).isEqualTo(message.obj.number)
         assertThat(it.obj.text).isEqualTo(message.obj.text)
         assertThat(context[propertyKey].toString()).isEqualToIgnoringCase(propertyValue)
         countDownLatch.countDown()
         it.obj.text
      }
      event
         .messageBuilder()
         .property(propertyKey, propertyValue)
         .sendAndReceive(message)
         .thenAccept {
            assertThat(it).isEqualTo(message.obj.text)
         }

      Await(countDownLatch)
   }

   @Test
   fun sendReply_parallel_two() {
      val event1 = newEvent<String, Int>()
      val event2 = newEvent<Int, String>()
      val asString = "12345"
      val asInt = 12345
      val sentTimesEach = 100
      val countDownLatch = CountDownLatch(sentTimesEach * 2)

      event1.replier { it -> it.toInt() }
      event2.replier { it -> it.toString() }

      for(i in 1..sentTimesEach) {
         CompletableFuture.runAsync {
            event2.sendAndReceive(asInt).thenAccept {
               assertThat(it).isEqualTo(asString)
               countDownLatch.countDown()
            }
            event1.sendAndReceive(asString).thenAccept {
               assertThat(it).isEqualTo(asInt)
               countDownLatch.countDown()
            }
         }
      }

      Await(countDownLatch)
   }

   @Test
   fun sendReply_parallel_single() {
      val event = newEvent<Int, String>()
      val startValue = 1
      val sentTimes = 1000
      val countDownLatch = CountDownLatch(sentTimes)

      event.replier { it -> it.toString() }

      for(i in 1..sentTimes) {
         CompletableFuture.runAsync {
            event.sendAndReceive(startValue + i).thenAccept {
               assertThat(it).isEqualTo((startValue + i).toString())
               countDownLatch.countDown()
            }
         }
      }

      Await(countDownLatch)
   }

}